### PR TITLE
Genbank refactor

### PR DESCRIPTION
**Pull request by JamesJeffryes**
There's a lot here:
1.Simplified build by adding local WS client, and removing transform, data_api and genome_annotation_api dependencies
2. Added common functions genome interface including taxon lookup, genome_tier assignment and genome validation.
3. GenbankUploaderScript.py is now entirely replaced with GenbankToGenome
4. Total refactor on the downloader once again using Biopython for formatting
5. Consolidated unit tests and added coverage (~90%) for genbank functions

Currently saves to NewGenome type (but downloader tests old genome type too) so you will have to test against CI

### CODE REVIEW

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 1]
```diff
@@ -177,50 +163,391 @@ def stage_input(self, params):
...
+        print("Feature Counts: ", genome['feature_counts'])
+        return genome
+
+    def _save_assembly(self, contigs, params):
+        print("Saving sequence as Assembly object")
+        assembly_id = "{}_assembly".format(params['genome_name'])
+        fasta_file = "{}/{}_assembly.fasta".format(
+            self.cfg.sharedFolder, params['genome_name'], self.time_string)
+        Bio.SeqIO.write(contigs, fasta_file, "fasta")
+        assembly_ref = self.aUtil.save_assembly_from_fasta(
```
[the remark this answers is not in the document]
Comment (JamesJeffryes) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: No need in this case

[the remark this answers is not in the document]
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: The "Definition" line from the GenBank file should be used as the description.

File: lib/GenomeFileUtil/core/GenomeInterface.py [hunk 2]
```diff
@@ -141,13 +148,95 @@ def save_one_genome(self, params):
...
+        if genome['domain'] == "Bacteria" and len(genome.get('cdss', [])) \
+                != len(genome['features']):
+            warnings.append("For prokaryotes, CDS array should be the same "
+                            "length as the Features array.")
+        if len(genome.get('mrnas', [])) != len(genome.get('cdss', [])):
+            warnings.append("mRNA array should be the same length as the CDS"
+                            "array if present.")
+        if "molecule_type" in genome and genome['molecule_type'] != "DNA":
+            warnings.append("Genome molecule_type is expected to be DNA but is"
+                            "actually " + str(genome['molecule_type']))
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenomeInterface.py]: If molecule type is "DNA" or "ds-DNA" everything is fine. If the molecule type is "ss-RNA","ds-RNA" or "ss-DNA", then we need to confirm that the domain is either Virus or Viroid. If it is not we need to do a warning/error. If it is none of those we need to do a warning/error.  The behavior will probably be determined by running in strict or loose mode.

File: lib/GenomeFileUtil/core/GenomeInterface.py [hunk 3]
```diff
@@ -141,13 +148,95 @@ def save_one_genome(self, params):
...
+        log('Validating genome object contents')
+        warnings = []
+        if len(genome.get('cdss', [])) < len(genome['features']):
+            warnings.append("CDS array should be at at least as long as the "
+                            "Features array.")
+        if genome['domain'] == "Bacteria" and len(genome.get('cdss', [])) \
+                != len(genome['features']):
+            warnings.append("For prokaryotes, CDS array should be the same "
+                            "length as the Features array.")
+        if len(genome.get('mrnas', [])) != len(genome.get('cdss', [])):
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenomeInterface.py]: Is this not just checking the size vs one another. Should only be a warning if mRNAs are present.

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 4]
```diff
@@ -177,50 +163,391 @@ def stage_input(self, params):
...
+        contigs = Bio.SeqIO.parse(file_path, "genbank")
+        assembly_ref = self._save_assembly(contigs, params)
+        assembly_data = self.dfu.get_objects(
+            {'object_refs': [assembly_ref],
+             'ignore_errors': 0})['data'][0]['data']
+        genome = {
+            "id": params['genome_name'],
+            "type": params['type'],
+            "original_source_file_name": os.path.basename(file_path),
+            "genetic_code": params['genetic_code'],
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenbankToGenome.py]: If this is not passed in you need to get the Genetic code from the Taxonomy object.

File: lib/GenomeFileUtil/core/GenbankToGenome.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: See detailed information where save_assembly is called.

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 5]
```diff
@@ -177,50 +163,391 @@ def stage_input(self, params):
...
+                    begin = int(part.end)
+                loc.append((
+                        record.id,
+                        begin + 1,
+                        strand_trans[part.strand],
+                        len(part)))
+            return loc
+
+        def _aliases(feat):
+            keys = ('locus_tag', 'old_locus_tag', 'protein_id')
```
[the remark this answers is not in the document]
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: Also EC_number
Ex: /EC_number="2.3.1.179"

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 6]
```diff
@@ -177,50 +163,391 @@ def stage_input(self, params):
...
+                    return loc2[1] >= loc1[1] and (
+                    loc2[1] + loc2[3] <= loc1[1] + loc1[3])
+                else:
+                    return loc2[1] <= loc1[1] and (
+                    loc2[1] - loc2[3] >= loc1[1] - loc1[3])
+
+            # if any location in the CDS in contained by any location of the gene
+            return any([_contains(x, y) for x, y in
+                        itertools.product(feat1['location'], feat2['location'])])
+
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenbankToGenome.py]: This logic only works for Genes as a parent (and only if NOT transpliced). 
If Transpliced we are kind of stuck going positionally in the file.
We need much more involved logic to check mRNA parentage. Information in the PD.

File: lib/GenomeFileUtil/core/GenomeInterface.py (hunk 2 above)
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenomeInterface.py]: Also more checks in the PD, but they can be added in a later iteration.

File: lib/GenomeFileUtil/core/GenbankToGenome.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: I do not see the Definition line being passed as the contig description.

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 7]
```diff
@@ -411,25 +426,22 @@ def _contains(loc1, loc2):
...
                 len(gene["protein_translation"]) <
                     len(cds["protein_translation"])):
                 gene["protein_translation"] = cds["protein_translation"]
                 gene["protein_translation_length"] = len(
                     cds["protein_translation"])
-            # Merge cds["aliases"] -> gene["aliases"]
-            gene["aliases"] = list(set(cds.get('aliases', []) +
-                                       gene.get('aliases', [])))
+            # Merge cds list attributes with gene
+            for key in ('function', 'aliases', 'db_xref', 'function'):
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenbankToGenome.py]: Have function twice

File: lib/GenomeFileUtil/core/GenbankToGenome.py (hunk 6 above)
Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: What is the plan for CDS and mRNA relationships?

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 8]
```diff
@@ -177,50 +163,427 @@ def stage_input(self, params):
...
+                    begin = int(part.end)
+                loc.append((
+                        record.id,
+                        begin + 1,
+                        strand_trans[part.strand],
+                        len(part)))
+            return loc
+
+        def _aliases(feat):
+            keys = ('locus_tag', 'old_locus_tag', 'protein_id',
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenbankToGenome.py]: We "EC_number" added here as well.  Kind of tricky. It is more of a db_xref, but we want it to go back to its own key value pair in the GB download ex: /EC_number="1.6.99.3"

Comment (jkbaumohl) [reply lib/GenomeFileUtil/core/GenbankToGenome.py]: Not Addressed

File: lib/GenomeFileUtil/core/GenbankToGenome.py [hunk 9]
```diff
@@ -177,50 +163,427 @@ def stage_input(self, params):
...
+                    'transcript_id', 'gene')
```
Comment (jkbaumohl) [inline lib/GenomeFileUtil/core/GenbankToGenome.py]: Odd I do not see it here.  Let me refresh page.

**Pull request merged**